README.md
# Rust HDC2080, HDC2021 and HDC2010 Low-Power Humidity and Temperature Digital Sensor Driver

[![crates.io](https://img.shields.io/crates/v/hdc20xx.svg)](https://crates.io/crates/hdc20xx)
[![Docs](https://docs.rs/hdc20xx/badge.svg)](https://docs.rs/hdc20xx)
[![Build Status](https://travis-ci.com/eldruin/hdc20xx-rs.svg?branch=master)](https://travis-ci.com/eldruin/hdc20xx-rs)
[![Coverage Status](https://coveralls.io/repos/github/eldruin/hdc20xx-rs/badge.svg?branch=master)](https://coveralls.io/github/eldruin/hdc20xx-rs?branch=master)

This is a platform agnostic Rust driver for the HDC2080, HDC2021 and HDC2010
low-power humidity and temperature digital sensor using the [`embedded-hal`]
traits.

This driver allows you to:
- Set the measurement mode. Temperature only or temperature and humidity. See: `set_measurement_mode()`.
- Make one shot measurement. See: `read()`.
- Read the data and interrupt status. See: `status()`.
- Trigger a software reset. See: `software_reset()`.
- Read the manufacturer ID. See: `manufacturer_id()`.
- Read the device ID. See: `device_id()`.

<!-- TODO
[Introductory blog post]()
-->

## The devices

The HDC2080 device is an integrated humidity and temperature sensor that
provides high accuracy measurements with very low power consumption in a
small DFN package. The capacitive-based sensor includes new integrated
digital features and a heating element to dissipate condensation and moisture.

The HDC2080 digital features include programmable interrupt thresholds to
provide alerts and system wake-ups without requiring a microcontroller to
be continuously monitoring the system. Combined with programmable sampling
intervals, a low power consumption, and a support for a 1.8-V supply voltage,
the HDC2080 is designed for battery-operated systems.

This driver is compatible with HDC2080, HDC2021 and HDC2010.

Datasheets: [HDC2080](https://www.ti.com/lit/ds/symlink/hdc2080.pdf), [HDC2021](https://www.ti.com/lit/ds/symlink/hdc2021.pdf), [HDC2010](https://www.ti.com/lit/ds/symlink/hdc2010.pdf)

## Usage

To use this driver, import this crate and an `embedded_hal` implementation,
then instantiate the device.

Please find additional examples using hardware in this repository: [driver-examples]

[driver-examples]: https://github.com/eldruin/driver-examples

```rust
use embedded_hal::blocking::delay::DelayMs;
use hdc20xx::{Hdc20xx, SlaveAddr};
use linux_embedded_hal::{Delay, I2cdev};

fn main() {
    let mut delay = Delay {};
    let dev = I2cdev::new("/dev/i2c-1").unwrap();
    let address = SlaveAddr::default();
    let mut sensor = Hdc20xx::new(dev, address);
    loop {
        loop {
            let result = sensor.read();
            match result {
                Err(nb::Error::WouldBlock) => delay.delay_ms(100_u8),
                Err(e) => {
                    println!("Error! {:?}", e);
                }
                Ok(data) => {
                    println!(
                        "Temperature: {:2}°C, Humidity: {:2}%",
                        data.temperature,
                        data.humidity.unwrap()
                    );
                }
            }
        }
    }
}
```

## Support

For questions, issues, feature requests, and other changes, please file an
[issue in the github project](https://github.com/eldruin/hdc20xx-rs/issues).

## License

Licensed under either of

 * Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or
   http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or
   http://opensource.org/licenses/MIT)

at your option.

### Contributing

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall
be dual licensed as above, without any additional terms or conditions.

[`embedded-hal`]: https://github.com/rust-embedded/embedded-hal


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Fixed
- Replaced usage of `nb::block!` macro in examples with explicit loop including
  a delay between repetitions. Without this the repetitions can be too quick and
  some devices do not respond correctly. Thanks to @bernardoaraujor for noticing this.

## [0.1.0] - 2020-03-01

Initial release to crates.io.

[Unreleased]: https://github.com/eldruin/hdc20xx-rs/compare/v0.1.0...HEAD
[0.1.0]: https://github.com/eldruin/hdc20xx-rs/releases/tag/v0.1.0


tests/status.rs
mod common;
use crate::common::{destroy, new, BitFlags as BF, Register, BASE_ADDR};
use embedded_hal_mock::i2c::Transaction as I2cTrans;
use hdc20xx::Status;

macro_rules! test {
    ($name:ident, $value:expr, $drdy:expr, $htemp:expr, $ltemp:expr,
     $hrh:expr, $lrh:expr) => {
        #[test]
        fn $name() {
            let mut sensor = new(&[I2cTrans::write_read(
                BASE_ADDR,
                vec![Register::DRDY],
                vec![$value],
            )]);
            let st = sensor.status().unwrap();
            assert_eq!(
                st,
                Status {
                    data_ready: $drdy,
                    high_temp_threshold_exceeded: $htemp,
                    low_temp_threshold_exceeded: $ltemp,
                    high_humidity_threshold_exceeded: $hrh,
                    low_humidity_threshold_exceeded: $lrh,
                }
            );
            destroy(sensor);
        }
    };
}

test!(all_false, 0, false, false, false, false, false);
test!(drdy, BF::DRDY_STATUS, true, false, false, false, false);
test!(htemp, BF::TH_STATUS, false, true, false, false, false);
test!(ltemp, BF::TL_STATUS, false, false, true, false, false);
test!(hrh, BF::HH_STATUS, false, false, false, true, false);
test!(lrh, BF::HL_STATUS, false, false, false, false, true);
test!(
    all_true,
    BF::DRDY_STATUS | BF::TH_STATUS | BF::TL_STATUS | BF::HH_STATUS | BF::HL_STATUS,
    true,
    true,
    true,
    true,
    true
);


tests/common/mod.rs
use embedded_hal_mock::i2c::{Mock as I2cMock, Transaction as I2cTrans};
use hdc20xx::{mode, Hdc20xx, SlaveAddr};

pub struct Register;
#[allow(unused)]
impl Register {
    pub const TEMP_L: u8 = 0x00;
    pub const HUMIDITY_L: u8 = 0x02;
    pub const DRDY: u8 = 0x04;
    pub const MEAS_CONF: u8 = 0x0F;
    pub const MANUFACTURER_ID_L: u8 = 0xFC;
    pub const DEVICE_ID_L: u8 = 0xFE;
}

pub struct BitFlags;
#[allow(unused)]
impl BitFlags {
    pub const SOFT_RESET: u8 = 1 << 7;
    pub const TEMP_ONLY: u8 = 1 << 1;
    pub const MEAS_TRIG: u8 = 1;
    pub const DRDY_STATUS: u8 = 1 << 7;
    pub const TH_STATUS: u8 = 1 << 6;
    pub const TL_STATUS: u8 = 1 << 5;
    pub const HH_STATUS: u8 = 1 << 4;
    pub const HL_STATUS: u8 = 1 << 3;
}
pub const BASE_ADDR: u8 = 0x40;

#[allow(unused)]
pub fn new(transactions: &[I2cTrans]) -> Hdc20xx<I2cMock, mode::OneShot> {
    Hdc20xx::new(I2cMock::new(transactions), SlaveAddr::default())
}

#[allow(unused)]
pub fn destroy<MODE>(sensor: Hdc20xx<I2cMock, MODE>) {
    sensor.destroy().done();
}


tests/integration.rs
mod common;
use crate::common::{destroy, new, BitFlags as BF, Register, BASE_ADDR};
use embedded_hal_mock::i2c::Transaction as I2cTrans;
use hdc20xx::MeasurementMode;

#[test]
fn can_create_and_destroy() {
    let sensor = new(&[]);
    destroy(sensor);
}

#[test]
fn can_get_device_id() {
    let dev_id = 0xABCD;
    let mut sensor = new(&[I2cTrans::write_read(
        BASE_ADDR,
        vec![Register::DEVICE_ID_L],
        vec![0xCD, 0xAB],
    )]);
    let id = sensor.device_id().unwrap();
    assert_eq!(dev_id, id);
    destroy(sensor);
}

#[test]
fn can_get_manufacturer_id() {
    let manuf_id = 0xABCD;
    let mut sensor = new(&[I2cTrans::write_read(
        BASE_ADDR,
        vec![Register::MANUFACTURER_ID_L],
        vec![0xCD, 0xAB],
    )]);
    let id = sensor.manufacturer_id().unwrap();
    assert_eq!(manuf_id, id);
    destroy(sensor);
}

macro_rules! set_test {
    ($name:ident, $method:ident, $reg:ident, $value:expr $(, $arg:expr)*) => {
        #[test]
        fn $name() {
            let mut sensor = new(&[I2cTrans::write(BASE_ADDR, vec![Register::$reg, $value])]);
            sensor.$method($($arg),*).unwrap();
            destroy(sensor);
        }
    };
}

set_test!(sw_reset_one_shot, software_reset, MEAS_CONF, BF::SOFT_RESET);
set_test!(
    set_temp_and_humidity_mode,
    set_measurement_mode,
    MEAS_CONF,
    0,
    MeasurementMode::TemperatureAndHumidity
);

set_test!(
    set_temp_only_mode,
    set_measurement_mode,
    MEAS_CONF,
    BF::TEMP_ONLY,
    MeasurementMode::TemperatureOnly
);

#[test]
fn can_make_one_shot_measurement_temp_and_humidity() {
    let transactions = [
        I2cTrans::write(BASE_ADDR, vec![Register::MEAS_CONF, BF::MEAS_TRIG]),
        I2cTrans::write_read(BASE_ADDR, vec![Register::DRDY], vec![0]),
        I2cTrans::write_read(BASE_ADDR, vec![Register::DRDY], vec![BF::DRDY_STATUS]),
        I2cTrans::write_read(
            BASE_ADDR,
            vec![Register::TEMP_L],
            vec![0xD9, 0x64, 0xEC, 0x91],
        ),
    ];
    let mut sensor = new(&transactions);
    sensor.read().expect_err("should block");
    sensor.read().expect_err("should block");
    let data = sensor.read().unwrap();
    assert!(data.temperature < 25.5);
    assert!(data.temperature > 24.5);
    let rh = data.humidity.unwrap();
    assert!(rh < 57.5);
    assert!(rh > 56.5);
    assert!(data.status.data_ready);
    destroy(sensor);
}

#[test]
fn can_make_one_shot_measurement_temp_only() {
    let transactions = [
        I2cTrans::write(BASE_ADDR, vec![Register::MEAS_CONF, BF::TEMP_ONLY]),
        I2cTrans::write(
            BASE_ADDR,
            vec![Register::MEAS_CONF, BF::TEMP_ONLY | BF::MEAS_TRIG],
        ),
        I2cTrans::write_read(BASE_ADDR, vec![Register::DRDY], vec![0]),
        I2cTrans::write_read(BASE_ADDR, vec![Register::DRDY], vec![BF::DRDY_STATUS]),
        I2cTrans::write_read(BASE_ADDR, vec![Register::TEMP_L], vec![0xD9, 0x64]),
    ];
    let mut sensor = new(&transactions);
    sensor
        .set_measurement_mode(MeasurementMode::TemperatureOnly)
        .unwrap();
    sensor.read().expect_err("should block");
    sensor.read().expect_err("should block");
    let data = sensor.read().unwrap();
    assert!(data.temperature < 25.5);
    assert!(data.temperature > 24.5);
    assert!(data.humidity.is_none());
    assert!(data.status.data_ready);
    destroy(sensor);
}


src/lib.rs
//! This is a platform agnostic Rust driver for the HDC2080, HDC2021 and
//! HDC2010 low-power humidity and temperature digital sensor using
//! the [`embedded-hal`] traits.
//!
//! [`embedded-hal`]: https://github.com/rust-embedded/embedded-hal
//!
//! This driver allows you to:
//! - Set the measurement mode. Temperature only or temperature and humidity. See: [`set_measurement_mode()`].
//! - Make one shot measurement. See: [`read()`].
//! - Read the data and interrupt status. See: [`status()`].
//! - Trigger a software reset. See: [`software_reset()`].
//! - Read the manufacturer ID. See: [`manufacturer_id()`].
//! - Read the device ID. See: [`device_id()`].
//!
//! [`set_measurement_mode()`]: struct.Hdc20xx.html#method.set_measurement_mode
//! [`read()`]: struct.Hdc20xx.html#method.read
//! [`status()`]: struct.Hdc20xx.html#method.status
//! [`software_reset()`]: struct.Hdc20xx.html#method.software_reset
//! [`manufacturer_id()`]: struct.Hdc20xx.html#method.manufacturer_id
//! [`device_id()`]: struct.Hdc20xx.html#method.device_id
//!
//! <!-- TODO
//! [Introductory blog post](TODO)
//! -->
//!
//! ## The devices
//!
//! The HDC2080 device is an integrated humidity and temperature sensor that
//! provides high accuracy measurements with very low power consumption in a
//! small DFN package. The capacitive-based sensor includes new integrated
//! digital features and a heating element to dissipate condensation and moisture.
//!
//! The HDC2080 digital features include programmable interrupt thresholds to
//! provide alerts and system wake-ups without requiring a microcontroller to
//! be continuously monitoring the system. Combined with programmable sampling
//! intervals, a low power consumption, and a support for a 1.8-V supply voltage,
//! the HDC2080 is designed for battery-operated systems.
//!
//! This driver is compatible with HDC2080, HDC2021 and HDC2010.
//!
//! Datasheets: [HDC2080](https://www.ti.com/lit/ds/symlink/hdc2080.pdf), [HDC2021](https://www.ti.com/lit/ds/symlink/hdc2021.pdf), [HDC2010](https://www.ti.com/lit/ds/symlink/hdc2010.pdf)
//!
//! ## Usage examples (see also examples folder)
//!
//! To use this driver, import this crate and an `embedded_hal` implementation,
//! then instantiate the device.
//!
//! Please find additional examples using hardware in this repository: [driver-examples]
//!
//! [driver-examples]: https://github.com/eldruin/driver-examples
//!
//! ### Make a one-shot temperature and humidity measurement
//!
//! ```no_run
//! use embedded_hal::blocking::delay::DelayMs;
//! use hdc20xx::{Hdc20xx, SlaveAddr};
//! use linux_embedded_hal::{Delay, I2cdev};
//!
//! let mut delay = Delay {};
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let address = SlaveAddr::default();
//! let mut sensor = Hdc20xx::new(dev, address);
//! loop {
//!     loop {
//!         let result = sensor.read();
//!         match result {
//!             Err(nb::Error::WouldBlock) => delay.delay_ms(100_u8),
//!             Err(e) => {
//!                 println!("Error! {:?}", e);
//!             }
//!             Ok(data) => {
//!                 println!(
//!                     "Temperature: {:2}°C, Humidity: {:2}%",
//!                     data.temperature,
//!                     data.humidity.unwrap()
//!                 );
//!             }
//!         }
//!     }
//! }
//! ```
//!
//! ### Use an alternative address
//!
//! ```no_run
//! use hdc20xx::{Hdc20xx, SlaveAddr};
//! use linux_embedded_hal::I2cdev;
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let address = SlaveAddr::Alternative(true);
//! let sensor = Hdc20xx::new(dev, address);
//! ```
//!
//! ### Configure measuring only the temperature
//!
//! ```no_run
//! use hdc20xx::{Hdc20xx, MeasurementMode, SlaveAddr};
//! use linux_embedded_hal::I2cdev;
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let address = SlaveAddr::default();
//! let mut sensor = Hdc20xx::new(dev, address);
//! sensor.set_measurement_mode(MeasurementMode::TemperatureOnly).unwrap();
//! ```
//!
//! ### Read the manufacturer and device ID
//!
//! ```no_run
//! use hdc20xx::{Hdc20xx, SlaveAddr};
//! use linux_embedded_hal::I2cdev;
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let address = SlaveAddr::default();
//! let mut sensor = Hdc20xx::new(dev, address);
//! let manuf_id = sensor.manufacturer_id().unwrap();
//! let dev_id = sensor.device_id().unwrap();
//! println!(
//!     "Manufacturer ID: {}, Device ID: {}",
//!     manuf_id, dev_id
//! );
//! ```
//!
//! ### Read the data and interrupt status
//!
//! ```no_run
//! use hdc20xx::{Hdc20xx, SlaveAddr};
//! use linux_embedded_hal::I2cdev;
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let address = SlaveAddr::default();
//! let mut sensor = Hdc20xx::new(dev, address);
//! let status = sensor.status().unwrap();
//! println!("Status: {:?}", status);
//! ```
//!
#![deny(unsafe_code, missing_docs)]
#![no_std]

use core::marker::PhantomData;
mod device_impl;
mod types;
pub use crate::types::{Error, Measurement, MeasurementMode, SlaveAddr, Status};
mod register_address;
use crate::register_address::{BitFlags, Register, BASE_ADDR};

/// HDC2080, HDC2021 and HDC2010 device driver
#[derive(Debug)]
pub struct Hdc20xx<I2C, MODE> {
    i2c: I2C,
    address: u8,
    meas_config: Config,
    was_measurement_started: bool,
    _mode: PhantomData<MODE>,
}

#[derive(Debug, Default, Clone, Copy)]
struct Config {
    bits: u8,
}

/// Mode marker
pub mod mode {
    /// One shot measurement mode
    pub struct OneShot(());
    /// Continuous measurement mode
    pub struct Continuous(());
}

mod private {
    use super::mode;
    pub trait Sealed {}
    impl Sealed for mode::OneShot {}
    impl Sealed for mode::Continuous {}
}


src/types.rs
use crate::BASE_ADDR;

/// All possible errors in this crate
#[derive(Debug)]
pub enum Error<E> {
    /// I²C communication error
    I2C(E),
    /// Invalid input data provided
    InvalidInputData,
}

/// Measurement result
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Measurement {
    /// Temperature (°C)
    pub temperature: f32,
    /// Relative Humidity (%RH)
    ///
    /// Optionally read depending on the measurement configuration
    pub humidity: Option<f32>,
    /// Last status
    pub status: Status,
}

/// Status
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Status {
    /// Whether data is ready
    pub data_ready: bool,
    /// Whether the temperature high threshold was exceeded
    pub high_temp_threshold_exceeded: bool,
    /// Whether the temperature low threshold was exceeded
    pub low_temp_threshold_exceeded: bool,
    /// Whether the humidity high threshold was exceeded
    pub high_humidity_threshold_exceeded: bool,
    /// Whether the humidity low threshold was exceeded
    pub low_humidity_threshold_exceeded: bool,
}

/// Measurement mode
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum MeasurementMode {
    /// Temperature and humidity (default)
    TemperatureAndHumidity,
    /// Temperature only
    TemperatureOnly,
}

impl Default for MeasurementMode {
    fn default() -> Self {
        MeasurementMode::TemperatureAndHumidity
    }
}

/// Possible slave addresses
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum SlaveAddr {
    /// Default slave address
    Default,
    /// Alternative slave address providing bit value for the SDO pin
    Alternative(bool),
}

impl Default for SlaveAddr {
    /// Default slave address
    fn default() -> Self {
        SlaveAddr::Default
    }
}

impl SlaveAddr {
    pub(crate) fn addr(self) -> u8 {
        match self {
            SlaveAddr::Default => BASE_ADDR,
            SlaveAddr::Alternative(false) => BASE_ADDR,
            SlaveAddr::Alternative(true) => BASE_ADDR | 1,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::BASE_ADDR as ADDR;
    use super::{MeasurementMode, SlaveAddr};

    #[test]
    fn can_get_default_address() {
        let addr = SlaveAddr::default();
        assert_eq!(ADDR, addr.addr());
    }

    #[test]
    fn can_generate_alternative_addresses() {
        assert_eq!(ADDR, SlaveAddr::Alternative(false).addr());
        assert_eq!(ADDR | 1, SlaveAddr::Alternative(true).addr());
    }

    #[test]
    fn can_get_default_measurement_mode() {
        assert_eq!(
            MeasurementMode::TemperatureAndHumidity,
            MeasurementMode::default()
        );
    }
}


examples/linux.rs
use embedded_hal::blocking::delay::DelayMs;
use hdc20xx::{Hdc20xx, SlaveAddr};
use linux_embedded_hal::{Delay, I2cdev};

fn main() {
    let mut delay = Delay {};
    let dev = I2cdev::new("/dev/i2c-1").unwrap();
    let address = SlaveAddr::default();
    let mut sensor = Hdc20xx::new(dev, address);
    loop {
        loop {
            let result = sensor.read();
            match result {
                Err(nb::Error::WouldBlock) => delay.delay_ms(100_u8),
                Err(e) => {
                    println!("Error! {:?}", e);
                }
                Ok(data) => {
                    println!(
                        "Temperature: {:2}°C, Humidity: {:2}%",
                        data.temperature,
                        data.humidity.unwrap()
                    );
                }
            }
        }
    }
}


src/device_impl.rs
use crate::{
    mode, BitFlags, Config, Error, Hdc20xx, Measurement, MeasurementMode, Register, SlaveAddr,
    Status,
};
use core::marker::PhantomData;
use embedded_hal::blocking::i2c;

impl<I2C> Hdc20xx<I2C, mode::OneShot> {
    /// Create new instance of the device.
    pub fn new(i2c: I2C, address: SlaveAddr) -> Self {
        Hdc20xx {
            i2c,
            address: address.addr(),
            meas_config: Config { bits: 0 },
            was_measurement_started: false,
            _mode: PhantomData,
        }
    }
}

impl<I2C, MODE> Hdc20xx<I2C, MODE> {
    /// Destroy driver instance, return I2C bus.
    pub fn destroy(self) -> I2C {
        self.i2c
    }
}

impl<I2C, E, MODE> Hdc20xx<I2C, MODE>
where
    I2C: i2c::WriteRead<Error = E> + i2c::Write<Error = E>,
{
    /// Set measurement mode
    pub fn set_measurement_mode(&mut self, mode: MeasurementMode) -> Result<(), Error<E>> {
        let config = match mode {
            MeasurementMode::TemperatureAndHumidity => {
                self.meas_config.with_low(BitFlags::TEMP_ONLY)
            }
            MeasurementMode::TemperatureOnly => self.meas_config.with_high(BitFlags::TEMP_ONLY),
        };
        self.write_register(Register::MEAS_CONF, config.bits)?;
        self.meas_config = config;
        Ok(())
    }

    /// Read data and interrupt status
    pub fn status(&mut self) -> Result<Status, Error<E>> {
        let status = self.read_register(Register::DRDY)?;
        Ok(Status {
            data_ready: (status & BitFlags::DRDY_STATUS) != 0,
            high_temp_threshold_exceeded: (status & BitFlags::TH_STATUS) != 0,
            low_temp_threshold_exceeded: (status & BitFlags::TL_STATUS) != 0,
            high_humidity_threshold_exceeded: (status & BitFlags::HH_STATUS) != 0,
            low_humidity_threshold_exceeded: (status & BitFlags::HL_STATUS) != 0,
        })
    }

    /// Get device ID
    pub fn device_id(&mut self) -> Result<u16, Error<E>> {
        self.read_double_register(Register::DEVICE_ID_L)
    }

    /// Get manufacturer ID
    pub fn manufacturer_id(&mut self) -> Result<u16, Error<E>> {
        self.read_double_register(Register::MANUFACTURER_ID_L)
    }
}

impl<I2C, E> Hdc20xx<I2C, mode::OneShot>
where
    I2C: i2c::WriteRead<Error = E> + i2c::Write<Error = E>,
{
    /// Make measurement of temperature or temperature and humidity according
    /// to the configuration.
    ///
    /// Note that all status except the last one once data becomes available
    /// are discarded.
    pub fn read(&mut self) -> nb::Result<Measurement, Error<E>> {
        if self.was_measurement_started {
            let status = self.status()?;
            if status.data_ready {
                let include_humidity = !self.meas_config.is_high(BitFlags::TEMP_ONLY);
                let mut data = [0; 4];
                if include_humidity {
                    self.read_data(Register::TEMP_L, &mut data)?;
                } else {
                    self.read_data(Register::TEMP_L, &mut data[..2])?;
                }
                self.was_measurement_started = false;
                let temp_raw = u16::from(data[0]) | (u16::from(data[1]) << 8);
                let temp = f32::from(temp_raw) / 65536.0 * 165.0 - 40.0;
                if include_humidity {
                    let rh_raw = u16::from(data[2]) | (u16::from(data[3]) << 8);
                    let rh = f32::from(rh_raw) / 65536.0 * 100.0;
                    Ok(Measurement {
                        temperature: temp,
                        humidity: Some(rh),
                        status,
                    })
                } else {
                    Ok(Measurement {
                        temperature: temp,
                        humidity: None,
                        status,
                    })
                }
            } else {
                Err(nb::Error::WouldBlock)
            }
        } else {
            let meas_conf = self.meas_config.with_high(BitFlags::MEAS_TRIG);
            self.write_register(Register::MEAS_CONF, meas_conf.bits)?;
            self.was_measurement_started = true;
            Err(nb::Error::WouldBlock)
        }
    }

    /// Software reset
    pub fn software_reset(&mut self) -> Result<(), Error<E>> {
        let conf = self.meas_config.with_high(BitFlags::SOFT_RESET);
        self.write_register(Register::MEAS_CONF, conf.bits)
    }
}

impl Config {
    fn with_high(self, mask: u8) -> Self {
        Config {
            bits: self.bits | mask,
        }
    }
    fn with_low(self, mask: u8) -> Self {
        Config {
            bits: self.bits & !mask,
        }
    }
    fn is_high(self, mask: u8) -> bool {
        (self.bits & mask) != 0
    }
}


src/register_address.rs
use crate::{Error, Hdc20xx};
use embedded_hal::blocking::i2c;

pub const BASE_ADDR: u8 = 0x40;

pub struct Register;
impl Register {
    pub const TEMP_L: u8 = 0x00;
    pub const DRDY: u8 = 0x04;
    pub const MEAS_CONF: u8 = 0x0F;
    pub const MANUFACTURER_ID_L: u8 = 0xFC;
    pub const DEVICE_ID_L: u8 = 0xFE;
}

pub struct BitFlags;
impl BitFlags {
    pub const SOFT_RESET: u8 = 1 << 7;
    pub const TEMP_ONLY: u8 = 1 << 1;
    pub const MEAS_TRIG: u8 = 1;
    pub const DRDY_STATUS: u8 = 1 << 7;
    pub const TH_STATUS: u8 = 1 << 6;
    pub const TL_STATUS: u8 = 1 << 5;
    pub const HH_STATUS: u8 = 1 << 4;
    pub const HL_STATUS: u8 = 1 << 3;
}

impl<I2C, E, MODE> Hdc20xx<I2C, MODE>
where
    I2C: i2c::Write<Error = E>,
{
    pub(crate) fn write_register(&mut self, register: u8, data: u8) -> Result<(), Error<E>> {
        let payload: [u8; 2] = [register, data];
        let addr = self.address;
        self.i2c.write(addr, &payload).map_err(Error::I2C)
    }
}

impl<I2C, E, MODE> Hdc20xx<I2C, MODE>
where
    I2C: i2c::WriteRead<Error = E>,
{
    pub(crate) fn read_double_register(&mut self, register: u8) -> Result<u16, Error<E>> {
        let mut data = [0, 0];
        self.read_data(register, &mut data)
            .and(Ok(u16::from(data[0]) | (u16::from(data[1]) << 8)))
    }

    pub(crate) fn read_register(&mut self, register: u8) -> Result<u8, Error<E>> {
        let mut data = [0];
        self.read_data(register, &mut data).and(Ok(data[0]))
    }

    pub(crate) fn read_data(&mut self, register: u8, data: &mut [u8]) -> Result<(), Error<E>> {
        let addr = self.address;
        self.i2c
            .write_read(addr, &[register], data)
            .map_err(Error::I2C)
    }
}
